File: 1406-stone-game-iii/1406-stone-game-iii.c
#include <string.h>
#include <limits.h>

#define MAX(a, b) ((a) > (b) ? (a) : (b))

char* stoneGameIII(int* stoneValue, int stoneValueSize) {
    int n = stoneValueSize;
    // Rolling 4-slot buffer to bypass all dynamic heap allocations
    int dp[4] = {0}; 

    for (int i = n - 1; i >= 0; i--) {
        int currentIdx = i % 4;
        dp[currentIdx] = INT_MIN; // Native 32-bit minimum integer value
        int take = 0;

        for (int k = 0; k < 3 && i + k < n; k++) {
            take += stoneValue[i + k];
            dp[currentIdx] = MAX(dp[currentIdx], take - dp[(i + k + 1) % 4]);
        }
    }

    int relativeDiff = dp[0];
    if (relativeDiff > 0) return "Alice";
    if (relativeDiff < 0) return "Bob";
    return "Tie";
}

File: 1406-stone-game-iii/1406-stone-game-iii.py
class Solution:
    def stoneGameIII(self, stoneValue: list[int]) -> str:
        n = len(stoneValue)
        # Allocate flat list matching the size layout
        dp = [0] * (n + 1)

        for i in range(n - 1, -1, -1):
            dp[i] = float('-inf') # Handles negative infinity boundaries cleanly
            take = 0
            
            for k in range(3):
                if i + k < n:
                    take += stoneValue[i + k]
                    dp[i] = max(dp[i], take - dp[i + k + 1])
                    
        if dp[0] > 0: return "Alice"
        if dp[0] < 0: return "Bob"
        return "Tie"
